test/t8_geometry/t8_gtest_geometry_lagrange.cxx

```diff
@@ -0,0 +1,259 @@
+t8code is a C library to manage a collection (a forest) of multiple
+connected adaptive space-trees of general element classes in parallel.
+
+Copyright (C) 2023 the developers
+
+t8code is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 2 of the License, or
+(at your option) any later version.
+
+t8code is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with t8code; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+/** \file t8_gtest_geometry_lagrange.cxx
+* Provide tests to check the geometry mappings via Lagrange interpolation.
+* The tests are parametrized on the element type and the polynomial degree.
+*/
+
+#include <array>
+#include <cstdlib>
+#include <sstream>
+#include <vector>
+
+#include <gtest/gtest.h>
+#include <test/t8_gtest_macros.hxx>
+#include <t8_eclass.h>
+#include <t8_vec.h>
+#include <t8_element_cxx.hxx>
+#include <t8_cmesh.h>
+#include <t8_cmesh_vtk_writer.h>
+#include <t8_schemes/t8_default/t8_default_cxx.hxx>
+#include <t8_forest/t8_forest.h>
+#include <t8_cmesh/t8_cmesh_examples.h>
+#include <t8_geometry/t8_geometry_implementations/t8_geometry_lagrange.hxx>
+
+#define MAX_POLYNOMIAL_DEGREE 2
+
+/**
+ * Generate a random double precision number in [0, 1].
+ * 
+ * \return  Random number.
+ */
+double
+random_number ()
+{
+  return static_cast<double> (std::rand ()) / RAND_MAX;
+}
+
+/**
+ * Add random perturbation to a vector.
+ * 
+ * \param vec            Vector to be perturbed.
+ * \param max_amplitude  Value of the maximum perturbation.
+ * \return               Perturbed vector.
+ */
+std::vector<double>
+perturb (const std::vector<double> &vec, const double max_amplitude)
+{
+  std::vector<double> perturbed_vec;
+  perturbed_vec.reserve (vec.size ());
+  for (const auto &v : vec)
+    perturbed_vec.push_back (v + random_number () * max_amplitude);
+  return perturbed_vec;
+}
+
+/**
+ * Return true if two arrays are element-wise equal within a tolerance.
+ * \tparam U   Iterable container with size.
+ * \tparam V   Iterable container with size.
+ * \param a    First array.
+ * \param b    Second arary.
+ * \param tol  Absolute tolerance.
+ * \return     True if all the elements of \a a and \a b are close to each
+ *             other, false otherwise.
+ */
+template <typename U, typename V>
+bool
+allclose (const U &a, const V &b, double tol = 1e-8)
+{
+  size_t size_a = a.size ();
+  size_t size_b = b.size ();
+  if (size_a != size_b)
+    return false;
+  for (size_t i = 0; i < size_a; ++i)
+    if (fabs (a[i] - b[i]) > tol)
+      return false;
+  return true;
+}
```

Comment:
```suggestion
allclose (const U &a, const V &b, double tol = T8_PRECISION_SQRT_EPS)
{
  T8_ASSERT (a.size () == b.size ());
  for(size_t i : indices(a)) {
    if (fabs (a[i] - b[i]) > tol)
      return false;
  }
  return true;
}
```

Essentially the same, but I think this should assert if both containers have the same size and this avoids manually looping over a variable.

Reply:
> this should assert if both containers have the same size

I mimicked the behavior of [`allclose` from NumPy](https://numpy.org/doc/stable/reference/generated/numpy.allclose.html) when writing this function, but I agree that we should make a length check; I will do it.

In which header can I find the `indices`  function? If you want to iterate through multiple containers, you can use `std::views::zip`, but that's C++23.

Reply:
You are right, I just copied this from an answer in a forum and did not check where the `indices()` function is located. And it seems like it does not exist. So we can use your solution, but we should still assert for the same length of both arrays.